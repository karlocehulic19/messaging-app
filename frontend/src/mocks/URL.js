global.URL.createObjectURL = vi.fn(() => "mockedFilePath");
global.URL.revokeObjectURL = vi.fn(() => null);
