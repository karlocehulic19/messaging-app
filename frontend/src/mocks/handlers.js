import { http, HttpResponse } from "msw";
import { config } from "../Constants";

export const handlers = [
  http.post(`${config.url.BACKEND_URL}/login`, async ({ request }) => {
    const body = await request.json();
    if (!body.username || !body.password) {
      return new HttpResponse("Missing credentials", { status: 401 });
    }

    if (body.username != "someUsername" || body.password != "somePassword") {
      return HttpResponse.json(
        { messages: ["Username or password is incorrect"] },
        { status: 401 }
      );
    }
    return HttpResponse.json({ token: "randomJWTtoken" }, { status: 200 });
  }),
];
