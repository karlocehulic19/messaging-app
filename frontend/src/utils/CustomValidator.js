class CustomValidator {
  constructor() {
    this.validationChain = {};
  }

  isAllowed(field) {
    return field in this.validationChain;
  }

  addValidation(field, callback, message) {
    if (!(field in this.validationChain)) {
      this.validationChain[field] = [];
    }
    this.validationChain[field].push([callback, message]);

    return null;
  }

  async #validateField(field, value) {
    for (const [callback, message] of this.validationChain[field]) {
      if (!(await callback(value))) throw { field, message };
    }

    return;
  }

  async validate(formData) {
    const errorMessages = {};
    const promises = [];

    for (const field of Object.keys(this.validationChain)) {
      promises.push(this.#validateField(field, formData[field]));
    }

    await Promise.allSettled(promises).then((result) => {
      result
        .filter(({ status }) => status === "rejected")
        .forEach(({ reason }) => {
          errorMessages[reason.field] = reason.message;
        });
    });

    return errorMessages;
  }
}

export class ValidatorBuilder {
  constructor() {
    this.validator = new CustomValidator();
  }

  field(field) {
    if (!field || typeof field !== "string")
      throw new Error("Field argument must be a non empty string");
    this.currField = field;

    return this;
  }

  addRule(rule, message) {
    if (!this.currField) {
      throw new Error("Field name must be specified before adding a rule");
    }

    this.validator.addValidation(this.currField, rule, message);
    return this;
  }

  build() {
    return this.validator;
  }
}
