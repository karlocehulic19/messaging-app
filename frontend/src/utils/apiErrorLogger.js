export default async function apiErrorLogger(error) {
  let resJSON;
  if (
    error.response &&
    error.response.headers.get("Content-Type").includes("application/json")
  )
    resJSON = await error.response.json();
  if (resJSON) console.error(resJSON);
  console.error(error);

  return resJSON;
}
