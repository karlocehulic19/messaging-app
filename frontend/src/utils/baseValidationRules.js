import { isAscii } from "validator";
import { isRequired } from "./validationExpressions";
import isEmail from "validator/lib/isEmail";

export const usernameRules = (ValidationBuilder, field = "username") => {
  ValidationBuilder.field(field)
    .addRule(isRequired, "Username field is required")
    .addRule(isAscii, "Invalid characters are provided");
};

export const emailRules = (ValidationBuilder, field = "email") => {
  ValidationBuilder.field(field)
    .addRule(isRequired, "Email field is required")
    .addRule(isEmail, "Value provided must be valid email address");
};
