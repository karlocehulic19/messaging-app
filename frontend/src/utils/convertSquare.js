import { Jimp } from "jimp";

export default async function convertSquare(
  pictureBuffer,
  pictureType,
  size = 200
) {
  const picture = await Jimp.fromBuffer(pictureBuffer);
  if (size > picture.height || size > picture.width)
    throw new Error("Too big crop size");
  picture.cover({ w: size, h: size });
  return await picture.getBase64(pictureType);
}
