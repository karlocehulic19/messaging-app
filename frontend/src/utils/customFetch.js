import { config } from "../Constants";
import { merge } from "lodash";

export class ResponseError extends Error {
  constructor(message, response) {
    super(message);
    this.response = response;
  }
}

function getDefaultOptions() {
  const defaultOptions = { headers: { "Content-Type": "application/json" } };
  const token = localStorage.getItem("site");

  if (token) {
    defaultOptions.headers.Authorization = `Bearer ${token}`;
  }

  return defaultOptions;
}

export default async function customFetch(url, options) {
  const response = await fetch(
    `${config.url.BACKEND_URL}${url}`,
    merge(getDefaultOptions(), options)
  );
  if (!response.ok) {
    throw new ResponseError("Bad request error", response);
  }

  return response;
}
