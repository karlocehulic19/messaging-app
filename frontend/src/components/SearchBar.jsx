import { useCallback, useState, useRef } from "react";
import useDebounce from "../hooks/useDebounce";
import SearchCard from "./SearchCard";
import customFetch from "../utils/customFetch";
import apiErrorLogger from "../utils/apiErrorLogger";
import styles from "./styles/SearchBar.module.css";

export default function SearchBar() {
  const [search, setSearch] = useState("");
  const [users, setUsers] = useState([]);
  const [searchBarState, setSearchBarState] = useState("result");
  const [focused, setFocus] = useState(false);
  const searchBarRef = useRef();

  const isOpen =
    (users.length ||
      searchBarState == "loading" ||
      searchBarState == "notfound" ||
      searchBarState == "error") &&
    focused;

  if (!users.length && search && searchBarState == "result")
    setSearchBarState("notfound");
  else if (search == "" && searchBarState == "loading")
    setSearchBarState("result");
  useDebounce(
    useCallback(() => {
      if (search != "") {
        customFetch(`/users?s=${search}`)
          .then((response) => response.json())
          .then((users) => {
            setUsers(users);
            setSearchBarState("result");
          })
          .catch((error) => {
            setSearchBarState("error");
            apiErrorLogger(error);
          });
      }
    }, [search]),
    500,
    search
  );

  const handleSearchChange = (e) => {
    setSearch(e.target.value);
    setSearchBarState("loading");
  };

  const handleBlur = (e) => {
    if (!e.currentTarget.contains(e.relatedTarget)) {
      setFocus(false);
    }
  };

  return (
    <div
      tabIndex={-1}
      onFocus={() => setFocus(true)}
      onBlur={handleBlur}
      role="search"
      id={styles["user-search-bar"]}
      className={isOpen ? styles.open : null}
    >
      <input
        ref={searchBarRef}
        onChange={handleSearchChange}
        value={search}
        role="searchbox"
        type="text"
        name="searchbox"
        placeholder="Search for users by username..."
        id={styles.searchbar}
      />
      {focused && (
        <div id={styles["user-search-listings"]} aria-label="Found users">
          {searchBarState == "error" && <span>Ups! An error occurred!</span>}
          {searchBarState == "loading" && <span>Searching...</span>}
          {searchBarState == "notfound" && <span>No users found</span>}
          {searchBarState == "result" &&
            users.map((user) => (
              <SearchCard
                key={user.username}
                username={user.username}
                searchBarRef={searchBarRef}
              ></SearchCard>
            ))}
        </div>
      )}
    </div>
  );
}
