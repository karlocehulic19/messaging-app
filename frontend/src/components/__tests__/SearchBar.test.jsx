import { describe, expect, vi } from "vitest";
import { screen, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SearchBar from "../SearchBar";
import { server } from "../../mocks/node";
import * as customFetch from "../../utils/customFetch";
import { BrowserRouter } from "react-router-dom";

server.listen();

describe("<SearchBar />", () => {
  it("renders correctly", () => {
    render(<SearchBar />);
    expect(screen.getByRole("search")).toMatchSnapshot();
  });

  it("display possible search candidates", async () => {
    render(<SearchBar />, { wrapper: BrowserRouter });
    const user = userEvent.setup();

    const customFetchSpy = vi.spyOn(customFetch, "default");

    await user.click(screen.getByRole("searchbox"));
    await user.keyboard("t");

    await waitFor(() => expect(customFetchSpy).toBeCalledWith("/users?s=t"), {
      timeout: 2010,
    });

    const usersFetch = customFetchSpy.mock.calls.filter((call) =>
      call[0].includes("?s=")
    );

    expect(usersFetch.length).toBe(1);
    expect(screen.getByLabelText("Found users")).toMatchSnapshot();
  });
});
