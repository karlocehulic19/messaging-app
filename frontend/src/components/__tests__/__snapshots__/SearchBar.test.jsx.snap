// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchBar /> > display possible search candidates 1`] = `
<div
  aria-label="Found users"
  id="_listing-div_45bafb"
>
  <div
    aria-label="Test user"
    class="_search-card_1db872"
    tabindex="-1"
  >
    <img
      alt="Test profile picture"
      class="profile-picture _profile-picture_1db872"
      src="/src/assets/default-profile-picture.jpeg"
    />
    <span
      class="_username-span_1db872"
    >
      Test
    </span>
  </div>
  <div
    aria-label="Test2 user"
    class="_search-card_1db872"
    tabindex="-1"
  >
    <img
      alt="Test2 profile picture"
      class="profile-picture _profile-picture_1db872"
      src="/src/assets/default-profile-picture.jpeg"
    />
    <span
      class="_username-span_1db872"
    >
      Test2
    </span>
  </div>
  <div
    aria-label="Tim user"
    class="_search-card_1db872"
    tabindex="-1"
  >
    <img
      alt="Tim profile picture"
      class="profile-picture _profile-picture_1db872"
      src="/src/assets/default-profile-picture.jpeg"
    />
    <span
      class="_username-span_1db872"
    >
      Tim
    </span>
  </div>
</div>
`;

exports[`<SearchBar /> > display possible search candidates 2`] = `"_open_45bafb"`;

exports[`<SearchBar /> > displays error message on wrong failed requests 1`] = `
<div
  aria-label="Found users"
  id="_listing-div_45bafb"
>
  <span>
    Ups! An error occurred!
  </span>
</div>
`;

exports[`<SearchBar /> > displays error message on wrong failed requests 2`] = `"_open_45bafb"`;

exports[`<SearchBar /> > displays not found when no users are found 1`] = `"_open_45bafb"`;

exports[`<SearchBar /> > displays searching text before users are loaded 1`] = `
<div
  aria-label="Found users"
  id="_listing-div_45bafb"
>
  <span>
    Searching...
  </span>
</div>
`;

exports[`<SearchBar /> > displays searching text before users are loaded 2`] = `"_open_45bafb"`;

exports[`<SearchBar /> > displays searching text before users are loaded 3`] = `
<div
  aria-label="Found users"
  id="_listing-div_45bafb"
>
  <div
    aria-label="Test user"
    class="_search-card_1db872"
    tabindex="-1"
  >
    <img
      alt="Test profile picture"
      class="profile-picture _profile-picture_1db872"
      src="/src/assets/default-profile-picture.jpeg"
    />
    <span
      class="_username-span_1db872"
    >
      Test
    </span>
  </div>
  <div
    aria-label="Test2 user"
    class="_search-card_1db872"
    tabindex="-1"
  >
    <img
      alt="Test2 profile picture"
      class="profile-picture _profile-picture_1db872"
      src="/src/assets/default-profile-picture.jpeg"
    />
    <span
      class="_username-span_1db872"
    >
      Test2
    </span>
  </div>
  <div
    aria-label="Tim user"
    class="_search-card_1db872"
    tabindex="-1"
  >
    <img
      alt="Tim profile picture"
      class="profile-picture _profile-picture_1db872"
      src="/src/assets/default-profile-picture.jpeg"
    />
    <span
      class="_username-span_1db872"
    >
      Tim
    </span>
  </div>
</div>
`;

exports[`<SearchBar /> > renders correctly 1`] = `
<div
  id="_user-search-bar_45bafb"
  role="search"
  tabindex="-1"
>
  <input
    name="searchbox"
    role="searchbox"
    type="text"
    value=""
  />
</div>
`;
